File: Java Solutions/SlidingWindow/src/main/java/org/example/BestTimeBuySellStock.java
package org.example;

public class BestTimeBuySellStock {
    public static void main(String[] args) {
        BestTimeBuySellStock bestTimeBuySellStock = new BestTimeBuySellStock();
        int[] prices = {2, 1, 4};
        System.out.println(bestTimeBuySellStock.maxProfit(prices));
    }

    public int maxProfit(int[] prices) {
        int left = 0;
        int right = left + 1;
        int maxProfit = 0;
        while (right < prices.length) {
            int profit = prices[right] - prices[left];
            if (profit > maxProfit) {
                maxProfit = profit;
            }
            if (prices[left] > prices[right]) {
                left++;
                if (left == right) {
                    right++;
                }
                continue;
            }
            right++;
        }
        return maxProfit;
    }
}

File: Python Solutions/sliding_window/best_time_buy_sell_stock.py
from typing import List

def max_profit(prices: List[int]) -> int:
    max_profit_ = 0
    left = 0
    right = left+1
    while right < len(prices):
        profit = prices[right] - prices[left]
        if max_profit_ < profit:
            max_profit_ = profit
        if prices[left] > prices[right]:
            left += 1
            if left == right:
                right += 1
            continue
        right += 1
    return max_profit_

if __name__ == "__main__":
    prices = [2, 1, 4]
    print(max_profit(prices))